browser/ephemeral_storage/blob_url_browsertest.cc

```diff
@@ -0,0 +1,320 @@
+    }
+
+    // Expect blob created from a.com is available in iframe and vice versa.
+    EXPECT_EQ(std::to_string(0), FetchBlob(rfhs[3], a_com_registered_blobs[0]));
+    EXPECT_EQ(std::to_string(3), FetchBlob(rfhs[0], a_com_registered_blobs[3]));
+
+    // Expect blob created from b.com iframe is available in another b.com
+    // iframe and vice versa.
+    EXPECT_EQ(std::to_string(1), FetchBlob(rfhs[2], a_com_registered_blobs[1]));
+    EXPECT_EQ(std::to_string(2), FetchBlob(rfhs[1], a_com_registered_blobs[2]));
+  }
+
+  // Register blobs in b.com and its subframes, check they can be fetched from
+  // originating frames without any issues.
+  // Ensure no blobs from a.com are available to fetch in iframes.
+  std::vector<GURL> b_com_registered_blobs;
+  {
+    content::RenderFrameHost* main_frame =
+        LoadURLInNewTab(b_site_ephemeral_storage_url_)->GetMainFrame();
+    const std::vector<content::RenderFrameHost*> rfhs = GetFrames(main_frame);
+
+    for (size_t idx = 0; idx < rfhs.size(); ++idx) {
+      auto* rfh = rfhs[idx];
+      // Register blob.
+      b_com_registered_blobs.push_back(RegisterBlob(rfh, std::to_string(idx)));
+
+      // Blob should be fetchable from the same frame.
+      EXPECT_EQ(std::to_string(idx),
+                FetchBlob(rfh, b_com_registered_blobs[idx]));
+
+      // No blobs from a.com should be available.
+      for (size_t a_com_blob_idx = 0;
+           a_com_blob_idx < a_com_registered_blobs.size(); ++a_com_blob_idx) {
+        EXPECT_EQ("error",
+                  FetchBlob(rfh, a_com_registered_blobs[a_com_blob_idx]));
+      }
+    }
+  }
+
+  // Expect all a.com blobs (including the ones from 3p frames) are
+  // available in another a.com tab.
+  {
+    content::RenderFrameHost* main_frame =
+        LoadURLInNewTab(a_site_ephemeral_storage_url_)->GetMainFrame();
+    const std::vector<content::RenderFrameHost*> rfhs = GetFrames(main_frame);
+
+    for (size_t idx = 0; idx < rfhs.size(); ++idx) {
+      auto* rfh = rfhs[idx];
+      // All blobs from another a.com tab should be avilable.
+      EXPECT_EQ(std::to_string(idx),
+                FetchBlob(rfh, a_com_registered_blobs[idx]));
+
+      // No blobs from b.com should be available.
+      for (size_t b_com_blob_idx = 0;
+           b_com_blob_idx < b_com_registered_blobs.size(); ++b_com_blob_idx) {
+        EXPECT_EQ("error",
+                  FetchBlob(rfh, b_com_registered_blobs[b_com_blob_idx]));
+      }
+    }
+
+    // Close the first a.com tab, ensure all blobs created there become obsolete
+    // and can't be fetched.
+    ASSERT_TRUE(browser()->tab_strip_model()->CloseWebContentsAt(
+        1, TabStripModel::CloseTypes::CLOSE_NONE));
+    content::RunAllTasksUntilIdle();
+    for (size_t idx = 0; idx < rfhs.size(); ++idx) {
+      auto* rfh = rfhs[idx];
+      EXPECT_EQ("error", FetchBlob(rfh, a_com_registered_blobs[idx]));
+    }
+  }
+}
+
+// Tests for the blob: URL scheme, originally implemented in
```

Comment:
We're copying upstream tests here? That seems problematic to me because they will get out of sync